### PR TITLE
Support for Artifactory trigger

### PR DESCRIPTION
**Pull request by AlexeiVainshtein**

### CODE REVIEW

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                        logger.fine("Updating " + job.getName());
+                        project.save();
+                        project.scheduleBuild(new ArtifactoryCause(itemLastModified.getUri()));
+                        return;
+                    }
+
+                    if (job instanceof WorkflowJob) {
+                        WorkflowJob project = (WorkflowJob) job;
+                        project.save();
+                        logger.fine("Updating " + job.getName());
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: Let's move this log line to be before *project.save()*.

**Comment by eyalbe4** [comment]
Added a few questions and suggestions @AlexeiVainshtein.
LMK what you think.

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryCause.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    private String url;
+
+    public ArtifactoryCause(String url) {
+        this.url = url;
+    }
+
+    @Override
+    public String getShortDescription() {
+        return "The build was triggered by Artifactory trigger." +
+                "\nThe artifact that triggered the build is: " + url;
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryCause.java]: Done

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    public ServerDetails getDetails() {
+        return details;
+    }
+
+    @Override
+    public DescriptorImpl getDescriptor() {
+        return (DescriptorImpl) super.getDescriptor();
+    }
+
+    private long getLastModified(String date) throws ParseException {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: Done

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        if (job == null) {
+            return;
+        }
+
+        ArtifactoryServer artifactoryServer = RepositoriesUtils.getArtifactoryServer(details.getArtifactoryName(), RepositoriesUtils.getArtifactoryServers());
+        ArtifactoryHttpClient client = new ArtifactoryHttpClient(artifactoryServer.getUrl(), artifactoryServer.getDeployerCredentialsConfig().provideUsername(job), artifactoryServer.getDeployerCredentialsConfig().providePassword(job), new NullLog());
+        try {
+            ArtifactoryItemLastModified itemLastModified = client.getItemLastModified(path);
+            if (itemLastModified != null) {
+                long artifactoryResponseLastModified = getLastModified(itemLastModified.getLastModified());
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: Done

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java (hunk 1 above)
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: Done

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    public void run() {
+        if (job == null) {
+            return;
+        }
+
+        ArtifactoryServer artifactoryServer = RepositoriesUtils.getArtifactoryServer(details.getArtifactoryName(), RepositoriesUtils.getArtifactoryServers());
+        ArtifactoryHttpClient client = new ArtifactoryHttpClient(artifactoryServer.getUrl(), artifactoryServer.getDeployerCredentialsConfig().provideUsername(job), artifactoryServer.getDeployerCredentialsConfig().providePassword(job), new NullLog());
+        try {
+            ArtifactoryItemLastModified itemLastModified = client.getItemLastModified(path);
+            if (itemLastModified != null) {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: Done

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    private String path;
+    private ServerDetails details;
+    private long lastModified;
+
+    @DataBoundConstructor
+    public ArtifactoryTriggerWrapper(String path, String spec, ServerDetails details) throws ANTLRException {
+        super(spec);
+        this.path = path;
+        this.details = details;
+        this.lastModified = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: Done

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.logging.Logger;
+
+/**
+ * @author Alexei Vainshtein
+ */
+public class ArtifactoryTriggerWrapper extends SCMTrigger {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: I've changed it to Trigger and the class name as well

File: src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        this.lastModified = System.currentTimeMillis();
+    }
+
+    @Override
+    public void run() {
+        if (job == null) {
+            return;
+        }
+
+        ArtifactoryServer artifactoryServer = RepositoriesUtils.getArtifactoryServer(details.getArtifactoryName(), RepositoriesUtils.getArtifactoryServers());
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply src/main/java/org/jfrog/hudson/trigger/ArtifactoryTriggerWrapper.java]: I've added the check for null

**Pull request merged**